### PR TITLE
Clean up config and make formatter tokenizer setup reusable

### PR DESCRIPTION
**Pull request by sidpatelgit**
1. Clean up fine tuning config to match what we ended up with for the fine tuning run.
2. Pull some formatter and tokenizer utilities out so that we can reuse them for eval.
3. Eval work ahead

### CODE REVIEW

File: arc/architects.py [hunk 1]
```diff
@@ -154,3 +155,13 @@ def torch_call(self, examples):
                 if mid_pos < end_pos:
                     batch["labels"][i][beg_pos:mid_pos] = -100
         return batch
+
+
+def load_unsloth_4bit(model_path):
```
Comment (MinhxLe) [inline arc/architects.py:160]: nit: imo this function abstraction is more confusing than it is useful.

File: arc/config.py [hunk 2]
```diff
@@ -28,6 +31,14 @@ def __post_init__(self):
             raise ValueError("expected model_dtype to be None if load_in_4bit")
 
 
+@dataclass
+class FineTuningDataConfig:
+    train_dataset_path: str
+    eval_dataset_path: str
+    train_dataset_constructor: ta.Callable[[], Dataset]
```
Comment (MinhxLe) [inline arc/config.py:38]: nit: naming - functions should be as verb or actions. To be clear it is a function, you can add `_fn` at the end. This can be get_train_dataset().

File: arc/config.py [hunk 3]
```diff
@@ -28,6 +31,14 @@ def __post_init__(self):
             raise ValueError("expected model_dtype to be None if load_in_4bit")
 
 
+@dataclass
+class FineTuningDataConfig:
+    train_dataset_path: str
```
Comment (MinhxLe) [inline arc/config.py:36]: nit: naming. this is the "cached path" - else it's unclear to me why there is a path and also a function to generate a dataset.

another comment but not worth fixing for now - when thinking about what is configuration or not, we are making a call if it is something we want to pass control to a user. When thinking if you are caching a generated dataset or not, i think the only DoF (degree of freedom) is whether a cache is used or not - not the path.

File: arc/config.py [hunk 4]
```diff
@@ -111,42 +126,68 @@ def output_dir(self) -> str:
...
+    )
+    transformed_train_dataset = dst.concat(
+        base_train_dataset,
+        dst.apply_transform(base_train_dataset, t.Reflect(t.Reflect.Type.DIAGONAL)),
+        *[dst.apply_transform(base_train_dataset, t.Rotate(i)) for i in range(4)],
+        dst.apply_transform(base_train_dataset, t.PermuteColor(seed=42)),
+    )
+    return dst.concat(
+        transformed_train_dataset,
+        dst.shuffle_train_order(transformed_train_dataset, seed=42),
```
Comment (MinhxLe) [inline arc/config.py:146]: nit, seed should be taken in as a parameter maybe? and derived from the config.

File: arc/config.py [hunk 5]
```diff
@@ -167,18 +208,18 @@ def architects_model_and_tokenizer_preprocessor(model, tokenizer):
     ),
     sftt_config=FineTuningSFTTConfig(
         random_state=42,
-        max_seq_length=fmt_opts["max_tokens"],
+        max_seq_length=8192,
         data_collator_constructor=architects_data_collator_constructor,
```
Comment (MinhxLe) [inline arc/config.py:212]: is this something a developer actually configures or is how to take a dataset and transform into a collator given other configurations always the same?

File: arc/architects.py
Comment (MinhxLe) [inline arc/architects.py:1]: I can't comment on the exact line (8) but  wonder f we should just remove this confguration. it seems like a footgun that this configuration exists yet we have our own configuration.

File: arc/config.py [hunk 6]
```diff
@@ -111,42 +126,68 @@ def output_dir(self) -> str:
...
     name="architects",
     model_config=FineTuningModelConfig(
-        model="nvidia/Mistral-NeMo-Minitron-8B-Base",
+        model="chuanli11/Llama-3.2-3B-Instruct-uncensored",  # "nvidia/Mistral-NeMo-Minitron-8B-Base",
         model_dtype=None,
         load_in_4bit=True,
     ),
-    model_and_tokenizer_preprocessor=architects_model_and_tokenizer_preprocessor,
-    data_loader=architects_data_loader,
+    model_tokenizer_formatter_preprocessor=architects_model_tokenizer_formatter_preprocessor,
```
Comment (MinhxLe) [inline arc/config.py:184]: is this something we actually configure or is it always the same
1. initialize model, tokenizer given model name.
2. initialize formatter with its own configuration (head/tail token)
3. lobotomize tokenizer and model given formatter (not up to configuration)

Comment (sidpatelgit) [reply arc/config.py:184]: will move out of config

File: arc/config.py (hunk 5 above)
Comment (sidpatelgit) [reply arc/config.py:212]: will look to move this out of config

File: arc/architects.py
Comment (sidpatelgit) [reply arc/architects.py:1]: yes def let's remove

File: arc/architects.py (hunk 1 above)
Comment (sidpatelgit) [reply arc/architects.py]: got rid of it!

File: arc/config.py (hunk 3 above)
Comment (sidpatelgit) [reply arc/config.py]: changed to use_cache and putting path creation into the script

File: arc/config.py (hunk 2 above)
Comment (sidpatelgit) [reply arc/config.py]: done

File: arc/config.py (hunk 4 above)
Comment (sidpatelgit) [reply arc/config.py]: done, maybe a little hackily using a post_init in the data config

File: arc/config.py (hunk 5 above)
Comment (sidpatelgit) [reply arc/config.py]: done - i'm not super happy about importing architects stuff in the main fine tuning script, but fine for now until we want to change it up

File: arc/config.py (hunk 6 above)
Comment (sidpatelgit) [reply arc/config.py]: done
